src/Npgsql/Internal/Converters/RecordObjectArrayConverter.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Internal.Descriptors;
+using Npgsql.PostgresTypes;
+using Npgsql.Properties;
+
+namespace Npgsql.Internal.Converters;
+
+public class RecordObjectArrayConverter : PgStreamingConverter<object[]>
+{
+    readonly PgSerializerOptions _serializerOptions;
+
+    public RecordObjectArrayConverter(PgSerializerOptions serializerOptions)
+        => _serializerOptions = serializerOptions;
+
+    public override object[] Read(PgReader reader)
+        => Read(async: false, reader, CancellationToken.None).GetAwaiter().GetResult();
+
+    public override ValueTask<object[]> ReadAsync(PgReader reader, CancellationToken cancellationToken = default)
+        => Read(async: true, reader, cancellationToken);
+
+    async ValueTask<object[]> Read(bool async, PgReader reader, CancellationToken cancellationToken)
+    {
+        if (reader.ShouldBuffer(sizeof(int)))
+            await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+        var fieldCount = reader.ReadInt32();
+        var result = new object[fieldCount];
+
+        for (var i = 0; i < fieldCount; i++)
+        {
+            if (reader.ShouldBuffer(sizeof(uint) + sizeof(int)))
+                await reader.BufferData(async, sizeof(uint) + sizeof(int), cancellationToken).ConfigureAwait(false);
+
+            var typeOID = reader.ReadUInt32();
+            var fieldLen = reader.ReadInt32();
+
+            // Note that we leave .NET nulls in the object array rather than DBNull.
+            if (fieldLen == -1)
+                continue;
+
+            var postgresType =
+                _serializerOptions.TypeCatalog.GetPgType((Oid)typeOID).GetRepresentationalType()
+                ?? UnknownBackendType.Instance;
+            var typeInfo = _serializerOptions.GetDefaultTypeInfo(postgresType)
```

Comment:
Should we maybe add some internal extension methods?

```cs
GetxxOrThrow(bool isRead, ...) => Getxx(...) ?? isRead 
    ? throw new NotSupported("Reading is not supported ##") 
    : throw new NotSupported("Writing is not supported ##")
```

Reply:
I can see a lot of Get*TypeInfo overloads there, would make for lots of extensions :) I don't think it's too bad without it.

Reply:
Though we could do GetX vs. FindX as discussed offline if you feel it's useful enough